protocol/messenger_activity_center.go

```diff
@@ -318,11 +399,130 @@ func (m *Messenger) handleActivityCenterAccepted(state *ReceivedMessageState, me
 }
 
 func (m *Messenger) handleActivityCenterDismissed(state *ReceivedMessageState, message protobuf.SyncActivityCenterDismissed) error {
-	resp, err := m.DismissActivityCenterNotifications(context.TODO(), toHexBytes(message.Ids), false)
+	resp, err := m.DismissActivityCenterNotifications(context.TODO(), toHexBytes(message.Ids), message.Clock, false)
 
 	if err != nil {
 		return err
 	}
 
 	return state.Response.Merge(resp)
 }
+
+func (m *Messenger) handleSyncActivityCenterNotificationState(state *ReceivedMessageState, a *protobuf.SyncActivityCenterNotificationState) error {
+	s := &ActivityCenterState{
+		HasSeen:   a.HasSeen,
+		UpdatedAt: a.UpdatedAt,
+	}
+	n, err := m.persistence.UpdateActivityCenterNotificationState(s)
+	if err != nil {
+		return err
+	}
+	if n > 0 {
+		state.Response.SetActivityCenterState(s)
+	}
+	return nil
+}
+
+func (m *Messenger) handleSyncActivityCenterNotifications(state *ReceivedMessageState, a *protobuf.SyncActivityCenterNotifications) error {
+	var notifications []*ActivityCenterNotification
+	for _, n := range a.ActivityCenterNotifications {
+		notification, err := convertActivityCenterNotificationFromProtobuf(n)
+		if err != nil {
+			return err
+		}
+		affectedNum, err := m.persistence.SaveActivityCenterNotification(notification, false)
+		if err != nil {
+			return err
+		}
+		if affectedNum > 0 {
+			notifications = append(notifications, notification)
+		}
+	}
+	response, err := m.processActivityCenterNotifications(notifications, true)
+	if err != nil {
+		return err
+	}
+	return state.Response.Merge(response)
+}
+
+func convertActivityCenterNotificationToProtobuf(n *ActivityCenterNotification) (p *protobuf.SyncActivityCenterNotification, err error) {
+	if n == nil {
```

Comment:
This is a good example of a function I would prefer to receive a non-pointer value, because it means the compiler will guarantee `n` can't be nil, so we can remove this check. It leads to less defensive code.

Now callers of `convertActivityCenterNotificationToProtobuf` will (possibly) need to deal with the nil value, which is actually good because the *nil handling* will happen closer to where the problem is (similar to how error handling in Go is better handled closer to where it happened). This approach can also greatly help pinpoint issues in the system, because now we can log the unexpected nil issue right where it happened.

It's also interesting to accept a non-pointer value because the name of the function starts with `convert`, which strongly indicates `n` is not mutated.

This is a minor point @qfrank, and I'm aware not everyone agrees with it.

Reply:
Agree, great point!